nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/EnforceOrder.java

```diff
@@ -0,0 +1,538 @@
+        .name("skipped")
+        .description("A FlowFile that has an order younger than current, which means arrived too late and skipped, will be routed to this relationship.")
+        .build();
+
+    private final Set<Relationship> relationships;
+
+    public EnforceOrder() {
+        final Set<Relationship> rels = new HashSet<>();
+        rels.add(REL_SUCCESS);
+        rels.add(REL_WAIT);
+        rels.add(REL_OVERTOOK);
+        rels.add(REL_FAILURE);
+        rels.add(REL_SKIPPED);
+        relationships = Collections.unmodifiableSet(rels);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        descriptors.add(GROUP_IDENTIFIER);
+        descriptors.add(ORDER_ATTRIBUTE);
+        descriptors.add(INITIAL_ORDER);
+        descriptors.add(MAX_ORDER);
+        descriptors.add(BATCH_COUNT);
+        descriptors.add(WAIT_TIMEOUT);
+        descriptors.add(INACTIVE_TIMEOUT);
+        return descriptors;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        final List<ValidationResult> results = new ArrayList<>(super.customValidate(validationContext));
+
+        final Long waitTimeoutMillis = validationContext.getProperty(WAIT_TIMEOUT).asTimePeriod(TimeUnit.MICROSECONDS);
+        final Long inactiveTimeoutMillis = validationContext.getProperty(INACTIVE_TIMEOUT).asTimePeriod(TimeUnit.MICROSECONDS);
+
+        if (waitTimeoutMillis >= inactiveTimeoutMillis) {
+            results.add(new ValidationResult.Builder().input(validationContext.getProperty(INACTIVE_TIMEOUT).getValue())
+                    .subject(INACTIVE_TIMEOUT.getDisplayName())
+                    .explanation(String.format("%s should be longer than %s",
+                            INACTIVE_TIMEOUT.getDisplayName(), WAIT_TIMEOUT.getDisplayName()))
+                    .valid(false)
+                    .build());
+        }
+
+        return results;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+
+        final ComponentLog logger = getLogger();
+        final Integer batchCount = context.getProperty(BATCH_COUNT).asInteger();
+
+        final StateMap stateMap;
+        try {
+            stateMap = context.getStateManager().getState(Scope.LOCAL);
+        } catch (final IOException e) {
+            logger.error("Failed to retrieve state from StateManager due to {}" + e, e);
+            context.yield();
+            return;
+        }
+
+        final OrderingContext oc = new OrderingContext(context, session);
+
+        oc.groupStates.putAll(stateMap.toMap());
+
+        for (int i = 0; i < batchCount; i++) {
```

Comment:
It might be better to use session.get(batchCount) and then iterate over those, rather than individual calls to session.get()?